test/integration/transaction.test.js

```diff
@@ -414,73 +414,105 @@ if (current.dialect.supports.transactions) {
+            })
+            .catch(() => {});
+        };
 
-          await Promise.all([update(1, 0), update(0, 1)]);
+        it('should treat deadlocked transaction as rollback', async function () {
+          const Task = await getAndInitializeTaskModel(this.sequelize);
+
+          // This gets called twice simultaneously, and we expect at least one of the calls to encounter a
+          // deadlock (which effectively rolls back the active transaction).
+          // We only expect createTask() to insert rows if a transaction is active.  If deadlocks are handled
+          // properly, it should only execute a query if we're actually inside a real transaction.  If it does
+          // execute a query, we expect the newly-created rows to be destroyed when we forcibly rollback by
+          // throwing an error.
+          // tl;dr; This test is designed to ensure that this function never inserts and commits a new row.
+          await Promise.all([update(this.sequelize, Task, 1, 0), update(this.sequelize, Task, 0, 1)]);
 
           const count = await Task.count();
           // If we were actually inside a transaction when we called `Task.create({ id: 2 })`, no new rows should be added.
           expect(count).to.equal(2, 'transactions were fully rolled-back, and no new rows were added');
         });
+
+        it('should release the connection for a deadlocked transaction', async function () {
+          const Task = await getAndInitializeTaskModel(this.sequelize);
+
+          // 1 of 2 queries should deadlock and be rolled back by InnoDB
+          this.sinon.spy(this.sequelize.connectionManager, 'releaseConnection');
+          await Promise.all([update(this.sequelize, Task, 1, 0), update(this.sequelize, Task, 0, 1)]);
+
+          // Verify that both of the connections were released
+          expect(this.sequelize.connectionManager.releaseConnection.callCount).to.equal(2);
+
+          // Verify that a follow-up READ_COMMITTED works as expected.
+          // For unknown reasons, we need to explicitly rollback on MariaDB,
+          // even though the transaction should've automatically been rolled
+          // back.
+          // Otherwise, this READ_COMMITTED doesn't work as expected.
+          const User = this.sequelize.define('user', {
+            username: Support.Sequelize.STRING
+          });
+          await this.sequelize.sync({ force: true });
+          await this.sequelize.transaction(
+            { isolationLevel: Transaction.ISOLATION_LEVELS.READ_COMMITTED },
+            async transaction => {
+              const users0 = await User.findAll({ transaction });
+              await expect(users0).to.have.lengthOf(0);
+              await User.create({ username: 'jan' }); // Create a User outside of the transaction
+              const users = await User.findAll({ transaction });
+              return expect(users).to.have.lengthOf(1); // We SHOULD see the created user inside the transaction
```

Comment:
```suggestion
              await expect(users).to.have.lengthOf(1); // We SHOULD see the created user inside the transaction
```

Reply:
I actually removed the `await`s since `expect` in Chai is synchronous, so `expect` doesn't need to be returned or awaited. It makes the assertions a bit more readable too!

Reply:
@hsource Of course! Thank you!